### PR TITLE
Implemented coupon messaging

### LINKED ISSUE #2355: Coupons: coupon messaging
Part of #2158 
https://projects.invisionapp.com/share/NB7VM1GSG#/screens/214035463

### acceptance criteria:
- [ ] If a coupon applies to a course or a run, we should display a message regarding the coupon below the row in the dashboard, like in the mockup.
- [ ] If a coupon applies to a program, we should display the message in place of the financial aid card, which should disappear due to #2334

### PR DESCRIPTION
**Pull request by noisecapella**
#### What are the relevant tickets?
Fixes #2355 

#### What's this PR do?
Handles messages for telling user that coupons will be applied for a course or a program

#### How should this be manually tested?
Create two coupons, one for your program and one for a course that's available for purchase. (Course runs are not supported right now: #2440). You can attach one of these coupons by going to `/dashboard/?coupon=code_goes_here`. Attach the program one and you should see a card:
screenshot from 2017-01-27 16-21-15

then attach the course one and you should see a message below the course row:

screenshot from 2017-01-27 16-20-21

### CODE REVIEW

**Comment by aliceriot** [comment]
functionality looks great! took me a bit to figure out getting all the coupons created correctly. looking at the code now.

File: static/js/components/dashboard/CouponCard.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    const { coupon } = this.props;
+
+    return <Card shadow={0}>
+      <CardTitle>Coupons</CardTitle>
+      <div>
+        <CouponMessage coupon={coupon} />
+      </div>
+    </Card>;
+  }
+}
```
Comment (aliceriot) [inline static/js/components/dashboard/CouponCard.js]: this strikes me as a nice component to declare using an arrow function - kind of just a style choice though.

File: static/js/components/dashboard/CouponCard_test.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// @flow
+import React from 'react';
+import { shallow } from 'enzyme';
+import { assert } from 'chai';
+
+import CouponCard from './CouponCard';
+import { makeCoupon, makeProgram } from '../../factories/dashboard';
+
```
Comment (aliceriot) [inline static/js/components/dashboard/CouponCard_test.js]: extra blank line

File: static/js/components/dashboard/CouponMessage.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    if (coupon.content_type === COUPON_CONTENT_TYPE_PROGRAM) {
+      return ' for each course in this program';
+    } else if (coupon.content_type === COUPON_CONTENT_TYPE_COURSE ||
+      coupon.content_type === COUPON_CONTENT_TYPE_COURSERUN) {
+      return ' for this course';
+    } else {
+      return '';
+    }
+  };
```
Comment (aliceriot) [inline static/js/components/dashboard/CouponMessage.js]: I think it might be nice to implement this with a `Map`, mapping coupon types to their corresponding messages. You could declare that at the module level:

```js
const couponTargetMessages = new Map([
  [ COUPON_CONTENT_TYPE_PROGRAM, " for each course in this program" ],
  ...
]);
```

then this method could just look like this:

```js
makeCouponTargetMessage = () => {
  const { coupon } = this.props;
  
  return couponTargetMessages.get(coupon.content_type) || '';
}
```

just a suggestion!

File: static/js/containers/DashboardPage_test.js [hunk 4]
```diff
@@ -365,4 +368,50 @@ describe('DashboardPage', () => {
       });
     });
   });
+
+  describe('coupon messaging', () => {
+    let dashboard, coursePrices, availablePrograms, coupons, coupon, programId, availableProgram;
+    beforeEach(() => {
```
Comment (aliceriot) [inline static/js/containers/DashboardPage_test.js]: could you add a blank line above the `beforeEach` here?

File: static/js/global_init.js [hunk 5]
```diff
@@ -27,13 +27,8 @@ if (!Object.entries) {
 let jsdom = require('jsdom');
 require('jsdom-global')();
 
-let localStorageMock;
+let localStorageMock = require('./util/test_utils').localStorageMock;
```
Comment (aliceriot) [inline static/js/global_init.js]: what's the reason for this change?

File: static/js/lib/coupon.js [hunk 6]
```diff
@@ -89,3 +90,21 @@ export const _calculateRunPrice = (
...
 import { calculateRunPrice } from './coupon';
+
+export function makeAmountMessage(coupon: Coupon): string {
+  if (coupon.amount_type === COUPON_AMOUNT_TYPE_FIXED_DISCOUNT) {
+    return `$${coupon.amount} off`;
+  } else if (coupon.amount_type === COUPON_AMOUNT_TYPE_PERCENT_DISCOUNT) {
+    return `${coupon.amount.times(100).toDecimalPlaces(0).toString()}% off`;
+  } else {
+    return '';
+  }
```
Comment (aliceriot) [inline static/js/lib/coupon.js]: this might be a little more readable with a `switch` statement (debateable I guess, up to you!)

Comment (noisecapella) [inline static/js/lib/coupon.js]: Sure

File: static/js/global_init.js (hunk 5 above)
Comment (noisecapella) [inline static/js/global_init.js]: I was getting a bizarre timeout error when it was doing the lazy import. I moved it out here and it went away, so I figured ¯\\\_(ツ)\_/¯

Comment (aliceriot) [inline static/js/global_init.js]: oh weird. carry on then!

File: static/js/components/dashboard/CouponMessage.js (hunk 3 above)
Comment (noisecapella) [inline static/js/components/dashboard/CouponMessage.js]: To be honest this function is going away in the fixed price PR due to the messaging getting complicated, so I'll leave it as is for this one

File: static/js/components/CouponNotificationDialog.js [hunk 7]
```diff
@@ -40,43 +37,25 @@ const CouponNotificationDialog = (
...
-        of ${ amount.toString() } off</strong> the price
-        of <strong>each</strong> course in { programName }.
-      </p>;
-    }
+    title = `Coupon applied: ${makeAmountMessage(coupon)} each course!`;
+    message = <p>
+      This coupon gives <strong>a discount
+      of {makeAmountMessage(coupon)}</strong> the price
+      of <strong>each</strong> course in { programName }.
+    </p>;
```
Comment (noisecapella) [inline static/js/components/CouponNotificationDialog.js]: This is also going to change in the fixed price PR. In that PR I'm not sure there's enough that will get shared to make it worth sharing (though we can decide that in that PR)

File: static/js/containers/DashboardPage.js [hunk 8]
```diff
@@ -430,6 +432,13 @@ class DashboardPage extends React.Component {
         />;
       }
 
+      let couponCard;
+      // $FlowFixMe: thinks program is possibly undefined even though it's used above
```
Comment (noisecapella) [inline static/js/containers/DashboardPage.js]: I think I figured this out. It thinks that `program` could change to a null or undefined value by the time the callback is executed. If I define `programId` outside the callback the problem goes away

File: static/js/components/dashboard/CouponCard.js (hunk 1 above)
Comment (noisecapella) [inline static/js/components/dashboard/CouponCard.js]: Sure

**Comment by noisecapella** [comment]
Comments addressed

File: static/js/components/CouponNotificationDialog.js (hunk 7 above)
Comment (aliceriot) [inline static/js/components/CouponNotificationDialog.js]: sounds good
